src/account/UpgradeableModularAccount.sol

```diff
@@ -564,10 +562,58 @@ contract UpgradeableModularAccount is
     // solhint-disable-next-line no-empty-blocks
     function _authorizeUpgrade(address newImplementation) internal override {}
 
-    function _checkIfValidationApplies(bytes4 selector, FunctionReference validationFunction, bool isDefault)
-        internal
-        view
-    {
+    function _checkIfValidationAppliesCallData(
+        bytes calldata callData,
+        FunctionReference validationFunction,
+        bool isDefault
+    ) internal view {
+        bytes4 outerSelector = bytes4(callData[:4]);
+
+        _checkIfValidationAppliesSelector(outerSelector, validationFunction, isDefault);
+
+        if (outerSelector == IStandardExecutor.execute.selector) {
+            (address target,,) = abi.decode(callData[4:], (address, uint256, bytes));
+
+            if (target == address(this)) {
+                // There is no point to call `execute` to recurse exactly once - this is equivalent to just having
+                // the calldata as a top-level call.
+                revert SelfCallRecursionDepthExceeded();
+            }
+        } else if (outerSelector == IStandardExecutor.executeBatch.selector) {
+            // executeBatch may be used to batch account actions together, by targetting the account itself.
+            // If this is done, we must ensure all of the inner calls are allowed by the provided validation
+            // function.
+
+            (Call[] memory calls) = abi.decode(callData[4:], (Call[]));
+
+            for (uint256 i = 0; i < calls.length; ++i) {
+                if (calls[i].target == address(this)) {
+                    bytes4 nestedSelector = bytes4(calls[i].data);
+
+                    if (
+                        nestedSelector == IStandardExecutor.execute.selector
+                            || nestedSelector == IStandardExecutor.executeBatch.selector
+                    ) {
+                        // To prevent arbitrarily-deep recursive checking, we limit the depth of self-calls to one
+                        // for the purposes of batching.
+                        // This means that all self-calls must occur at the top level of the batch.
+                        // Note that plugins of other contracts using `executeWithAuthorization` may still
+                        // independently call into this account with a different validation function, allowing
+                        // composition of multiple batches.
+                        revert SelfCallRecursionDepthExceeded();
+                    }
+
+                    _checkIfValidationAppliesSelector(nestedSelector, validationFunction, isDefault);
+                }
+            }
+        }
```

Comment:
Should we explicitly check `executeWithAuthorization` and `executeUserOp` and only allow these to pass?

Reply:
- `executeUserOp` doesn't yet exist on this branch, but on the branch where it does, it requires `msg.sender` to be the entrypoint, so we don't need extra handling here.
- `executeWithAuthorization` will re-run a new validation function, so it should be OK to allow - it doesn't really make sense to call it in a nested way via `execute`/`executeBatch` due it just adding gas, but it won't break any security guarantees.

Reply:
Thanks! I meant reverting on [line 609](https://github.com/erc6900/reference-implementation/blob/5fda9a22e018c4450e0ff4fd18e7494b306a8a80/src/account/UpgradeableModularAccount.sol#L609) if random function selector is accidentally calling but not in (`executeWithAuthorization`, `executeUserOp`). Maybe it's a bit too defensive.